### ISSUE TITLE
Javascript - Go to definition does not work with variables (sublime does so VSC should too)

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#10110]
_From @darkguy2008 on July 28, 2016 17:16_

- VSCode Version: 1.3.1
- OS Version: Windows 8.1

Steps to Reproduce:

* Create a normal JS object (using ES6 syntax here):
```
class randomClass {
  constructor () {
    let self = this;
    this.variable = null;
    self.variable = 'lol';
  }
}
```

* Put the cursor in "variable" and press F12 (or click any of the definition stuff in the right-click menu - peek, go, etc.). Nothing happens, not even the tooltip shows anything.

- Expected behavior:
   VS Code should either:
   * Show a list of the found definitions in the command palette (Control+Shift+P) and when you scroll through them using the arrow keys the cursor/screen should scroll to where that definition is found (mimics Sublime's behavior which is great)
   * Show a list like in VS2015 when you click "Show all references" in a C# project. Same behavior: Click and it takes you where that definition is found.

Basically, there should be a way to find out where does a variable come from. A search works, or Control+D, but that way you can't go forward/backwards, whereas with the first proposed option you can do so :)

_Copied from original issue: Microsoft/vscode#9912_

**Comment by dbaeumer** [comment]
_From @kieferrm on July 28, 2016 18:28_

@darkguy2008 what would expect to happen?

**Comment by dbaeumer** [comment]
_From @darkguy2008 on July 28, 2016 18:55_

@kieferrm I have updated the issue accordingly, thanks!

**Comment by dbaeumer** [comment]
_From @darkguy2008 on August 2, 2016 20:22_

No updates about this yet? it does not need more info now, how to remove the label?

**Comment by dbaeumer** [comment]
_From @kieferrm on August 2, 2016 21:7_

When you cursor is at `self.varia|ble` you want to see that `this.variable` is the definition of it.

**Comment by dbaeumer** [comment]
_From @darkguy2008 on August 2, 2016 21:17_

@kieferrm Yeah of course! it does not matter if the cursor is in another function inside the same file, it should go where the (first?) definition is - or show a list where that variable is being defined. 

I understand it may be hard to have it work across different files, but Sublime works around it by just doing a full search of that variable and shows it in the command palette. While this is not exactly what one expects and sometimes it gets confused with files inside node_modules, it works and helps you to navigate through the code. For me, this is a blocker as I have to work with long JS files and I use this functionality a lot to figure out the variable declarations :)

**Comment by dbaeumer** [comment]
Moving to the TS team since this must come from the tsserver.

**Comment by mhegazy** [comment]
Should be working correctly in latest drop of TypeSript (`typescript@next`). VSCode ships with an older version of TypeScript. Please see [Using Newer TypeScript Versions](https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions) documentation for more details on updating your VSCode to use a different version of TypeScript.

**Comment by darkguy2008** [comment]
@mhegazy while this is a solution, it is not how it should be. VSCode should come with an updated typescript version automatically. Not all projects require (or even use) typescript in order to require it to be installed in the project and break it all. In my case, the project I'm working on does not use typescript and I could possibly get blamed because I'm including a dependency out of my own preferences of using a different editor.

The correct approach here would be to update VSCode's core typescript version instead. The proposed solution is a hassle for end users.

**Comment by mhegazy** [comment]
> @mhegazy while this is a solution, it is not how it should be. VSCode should come with an updated typescript version automatically.

this is what VSCode does. but it takes time. TS needs to release a new version, VSCode tests it, releases it in insider version, then in a stable release. if you want to try it out today, use the link above. You can also wait until VSCode version comes with the fix. up to you.

**Comment by darkguy2008** [comment]
@mhegazy well you're right, I didn't notice this issue was moved to typescript instead - this makes no sense. The idea of this issue was to tell VSCode project owners to do something about it (in this case, update their typescript version) but it got moved to the wrong project instead I guess :/
